cache/src/commonMain/kotlin/com/dropbox/kmp/external/cache3/LocalCache.kt

```diff
@@ -0,0 +1,1926 @@
+            get() {
+                for (e in accessQueue) {
+                    val weight = e.valueReference!!.weight
+                    if (weight > 0) {
+                        return e
+                    }
+                }
+                throw AssertionError()
+            }
+
+        /**
+         * Returns first entry of bin for given hash.
+         */
+        private fun getFirst(hash: Int): ReferenceEntry<K, V>? {
+            // read this volatile field only once
+            val table = table.value
+            return table[hash and table.size - 1]
+        }
+
+        // Specialized implementations of map methods
+        private fun getEntry(key: K, hash: Int): ReferenceEntry<K, V>? {
+            var e = getFirst(hash)
+            while (e != null) {
+                if (e.hash != hash) {
+                    e = e.next
+                    continue
+                }
+                val entryKey = e.key
+                if (key == entryKey) {
+                    return e
+                }
+                e = e.next
+            }
+            return null
+        }
+
+        private fun getLiveEntry(key: K, hash: Int, now: Long): ReferenceEntry<K, V>? {
+            val e = getEntry(key, hash)
+            if (e == null) {
+                return null
+            } else if (map.isExpired(e, now)) {
+                tryExpireEntries(now)
+                return null
+            }
+            return e
+        }
+
+        /**
+         * Gets the value from an entry. Returns null if the entry is invalid, partially-collected,
+         * loading, or expired.
+         */
+
+        fun get(key: K, hash: Int): V? {
+            return try {
+                if (count.value != 0) { // read-volatile
+                    val now = map.ticker()
+                    val e = getLiveEntry(key, hash, now) ?: return null
+                    val value = e.valueReference?.get()
+                    if (value != null) {
+                        recordRead(e, now)
+                        return value
+                    }
+                }
+                null
+            } finally {
+                postReadCleanup()
+            }
+        }
+
+        fun getOrPut(key: K, hash: Int, defaultValue: () -> V): V {
+            reentrantLock.lock()
+            return try {
+                if (count.value != 0) { // read-volatile
+                    val now = map.ticker()
+                    val e = getLiveEntry(key, hash, now)
+                    val value = e?.valueReference?.get()
+                    if (value != null) {
+                        recordRead(e, now)
+                        return value
+                    }
+                }
+                val default = defaultValue()
+                put(key, hash, default, false)
+                default
+            } finally {
+                reentrantLock.unlock()
+                postReadCleanup()
```

Comment:
Should this come before the `unlock` call above? I (think) in the call hierarchy we ultimately call `runLockedCleanup` which gets the lock again which feels like we could get into some race conditions.  Thoughts?